flight/Libraries/msp/msp.c

```diff
@@ -0,0 +1,263 @@
+/**
+ ******************************************************************************
+ * @file       msp.c
+ * @author     dRonin, http://dRonin.org/, Copyright (C) 2016
+ * @addtogroup Libraries
+ * @{
+ * @addtogroup MSP Protocol Library
+ * @{
+ * @brief Deals with packing/unpacking and MSP streams
+ *****************************************************************************/
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ * Additional note on redistribution: The copyright and license notices above
+ * must be maintained in each individual source file that is a derivative work
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+#include "msp.h"
+
+#define MSP_OVERHEAD_BYTES (2 /* preamble */ \
+                          + 1 /* direction */ \
+                          + 1 /* size */ \
+                          + 1 /* command */ \
+                          + 1 /* checksum */)
+
+enum msp_state {
+	MSP_STATE_IDLE,
+	MSP_STATE_PREAMBLE,
+	MSP_STATE_DIRECTION,
+	MSP_STATE_SIZE,
+	MSP_STATE_COMMAND,
+	MSP_STATE_DATA,
+	MSP_STATE_CHECKSUM,
+	MSP_STATE_DISCARD,
+};
+
+struct msp_parser {
+	msp_handler_t handler;
```

Comment:
Probably want a magic/sentinel value here so we can abort safely if somebody above us blows their stack.